Portal/src/Datahub.Infrastructure/Services/Cost/WorkspaceBudgetManagementService.cs

```diff
@@ -0,0 +1,246 @@
+                };
+                ctx.Project_Credits.Add(projectCredits);
+            }
+
+            var currentSpent = await GetWorkspaceBudgetSpentAsync(workspaceAcronym, budgetId);
+            var beforeUpdateBudgetSpent = (decimal)projectCredits.BudgetCurrentSpent;
+            projectCredits.BudgetCurrentSpent = (double)currentSpent;
+
+            ctx.Project_Credits.Update(projectCredits);
+            await ctx.SaveChangesAsync();
+            return beforeUpdateBudgetSpent;
+        }
+
+        /// <summary>
+        /// Gets the azure budget resource for the given budget id.
+        /// </summary>
+        /// <param name="budgetId">The budget identifier</param>
+        /// <returns>The consumption budget resource</returns>
+        /// <exception cref="Exception">Throws an exception if the budget with given id does not exist or if it could not be read</exception>
+        internal async Task<ConsumptionBudgetResource> GetBudgetAzureResource(string budgetId)
+        {
+            var budgetIdentifier = new ResourceIdentifier(budgetId);
+            var budget = _armClient.GetConsumptionBudgetResource(budgetIdentifier);
+            if (budget == null)
+            {
+                _logger.LogError($"Budget with id {budgetId} not found");
+                throw new Exception($"Budget with id {budgetId} not found");
+            }
+
+            var response = await budget.GetAsync();
+
+            if (response.Value == null)
+            {
+                _logger.LogError($"Budget with id {budgetId} not found");
+                throw new Exception($"Budget with id {budgetId} not found");
+            }
+
+            return response.Value;
+        }
+
+        /// <summary>
+        /// Gets the budget id for the given workspace acronym.
+        /// </summary>
+        /// <param name="workspaceAcronym">The workspace acronym</param>
+        /// <param name="rgName">Optional resource group name to use to get budget id</param>
+        /// <returns>The budget id</returns>
+        internal async Task<string> GetBudgetIdForWorkspace(string workspaceAcronym, string? rgName = null)
+        {
+            if (rgName is null)
+            {
+                using var ctx = await _dbContextFactory.CreateDbContextAsync();
+                var projectResources = ctx.Project_Resources2.Include(p => p.Project)
+                    .Where(p => p.Project.Project_Acronym_CD == workspaceAcronym).ToList();
+                var blobType = TerraformTemplate.GetTerraformServiceType(TerraformTemplate.AzureStorageBlob);
+                var blobResource = projectResources.First(r => r.ResourceType == blobType);
+                rgName = ParseResourceGroup(blobResource.JsonContent);
+            }
+
+            var budgetId =
+                $"/subscription/{_armClient.GetDefaultSubscription().Id.SubscriptionId}/resourceGroups/{rgName}/providers/Microsoft.Consumption/budgets/{rgName}-budget";
+            return budgetId;
```

Comment:
We'll need a TODO here as I'm going to be attaching the subscription id to a workspace (as there will be multiple subscriptions)